File: src/routes/warehouse.js
'use strict';
const express = require('express');
const { db } = require('../db');
const { requireLogin, requireRole } = require('../middleware/auth');
const { prijem, korekcia, stockState, AppError } = require('../services/stock');
const { otvoreneAlerty, posliSuhrn, exportCsv } = require('../services/alerts');

const r = express.Router();
r.use(requireLogin);

/* ---------------- SKLAD ---------------- */

/**
 * Sklad má jednu stránku so záložkami - stav, doobjednanie, príjem,
 * inventúra a pohyby. Menej položiek v menu, menej blúdenia.
 */
function zobrazSklad(req, res, zalozka) {
  const q = req.query.q ? String(req.query.q).trim() : '';
  const polozky = db.prepare(`SELECT i.*, c.name AS cat_name FROM items i
                              LEFT JOIN categories c ON c.id = i.category_id
                              WHERE i.active = 1
                                AND (? = '' OR i.code LIKE ? OR i.name LIKE ? OR i.location LIKE ?)
                              ORDER BY i.code`).all(q, `%${q}%`, `%${q}%`, `%${q}%`);
  for (const p of polozky) p.stav = stockState(p);

  const pohyby = db.prepare(`SELECT m.*, i.code, i.name, i.unit, u.full_name,
                                    rq.number AS ziadanka
                             FROM stock_moves m JOIN items i ON i.id = m.item_id
                             LEFT JOIN users u ON u.id = m.user_id
                             LEFT JOIN request_lines rl ON rl.id = m.request_line_id
                             LEFT JOIN requests rq ON rq.id = rl.request_id
                             ORDER BY m.id DESC LIMIT 100`).all();

  res.render('sklad', {
    zalozka, q, polozky, pohyby,
    alerty: otvoreneAlerty(),
    dodavatelia: db.prepare('SELECT id, name FROM suppliers WHERE active=1 ORDER BY name').all(),
  });
}

r.get('/sklad', requireRole('storekeeper', 'buyer'), (req, res) => {
  const povolene = ['stav', 'prijem', 'inventura', 'pohyby'];
  zobrazSklad(req, res, povolene.includes(req.query.tab) ? req.query.tab : 'stav');
});

r.get('/sklad/doobjednat', requireRole('storekeeper', 'buyer'),
  (req, res) => zobrazSklad(req, res, 'doobjednat'));

r.post('/sklad/prijem', requireRole('storekeeper'), (req, res, next) => {
  try {
    prijem({
      itemId: Number(req.body.item_id),
      qty: Number(req.body.qty),
      supplierId: req.body.supplier_id ? Number(req.body.supplier_id) : null,
      userId: req.session.user.id,
      note: req.body.note || null,
    });
    req.session.flash = 'Príjem zaevidovaný.';
  } catch (e) {
    if (!(e instanceof AppError)) return next(e);
    req.session.flash = e.message;
  }
  res.redirect(req.app.locals.base + '/sklad?tab=prijem');
});

r.post('/sklad/korekcia', requireRole('storekeeper'), (req, res, next) => {
  try {
    korekcia({
      itemId: Number(req.body.item_id),
      newQty: Number(req.body.new_qty),
      userId: req.session.user.id,
      note: req.body.note || null,
      type: req.body.typ === 'KOREKCIA' ? 'KOREKCIA' : 'INVENTURA',
    });
    req.session.flash = 'Stav upravený.';
  } catch (e) {
    if (!(e instanceof AppError)) return next(e);
    req.session.flash = e.message;
  }
  res.redirect(req.app.locals.base + '/sklad?tab=inventura');
});

r.post('/sklad/doobjednat/suhrn', requireRole('storekeeper', 'buyer'), (req, res) => {
  const n = posliSuhrn();
  req.session.flash = n ? `Podklad na doobjednanie odoslaný (${n} položiek).`
                        : 'Žiadne položky pod signálnou zásobou.';
  res.redirect(req.app.locals.base + '/sklad/doobjednat');
});

/** CSV pre oddelenie, ktoré vystavuje objednávky u dodávateľov. */
r.get('/sklad/doobjednat/export.csv', requireRole('storekeeper', 'buyer'), (req, res) => {
  const datum = new Date().toISOString().slice(0, 10);
  res.setHeader('Content-Type', 'text/csv; charset=utf-8');
  res.setHeader('Content-Disposition', `attachment; filename="doobjednat-${datum}.csv"`);
  res.send(exportCsv());
});

// Staré adresy nech nekončia na 404.
r.get('/upozornenia', (req, res) => res.redirect(req.app.locals.base + '/sklad/doobjednat'));
r.get('/sklad/karta/:id', (req, res) =>
  res.redirect(`${req.app.locals.base}/produkt/${req.params.id}`));

/* ---------------- DODAVATELIA ---------------- */

r.get('/dodavatelia', (req, res) => {
  const dodavatelia = db.prepare(`SELECT s.*,
      (SELECT COUNT(*) FROM supplier_items si WHERE si.supplier_id = s.id) AS pocet_poloziek
      FROM suppliers s WHERE s.active = 1 ORDER BY s.name`).all();
  res.render('dodavatelia', { dodavatelia });
});

r.get('/dodavatelia/:id', (req, res) => {
  const d = db.prepare('SELECT * FROM suppliers WHERE id=?').get(req.params.id);
  if (!d) return res.status(404).render('chyba', { msg: 'Dodávateľ neexistuje.' });
  const polozky = db.prepare(`SELECT i.*, si.supplier_sku, si.price, si.min_order_qty, si.is_primary
                              FROM supplier_items si JOIN items i ON i.id = si.item_id
                              WHERE si.supplier_id = ? ORDER BY i.code`).all(d.id);
  for (const p of polozky) p.stav = stockState(p);
  const ponuky = db.prepare(`SELECT q.*,
      (SELECT COUNT(*) FROM quote_lines ql WHERE ql.quote_id = q.id) AS riadkov
      FROM quotes q WHERE q.supplier_id = ? ORDER BY q.issued_at DESC`).all(d.id);
  res.render('dodavatel', { d, polozky, ponuky });
});

module.exports = r;

File: backend/app/routers/warehouse.py
"""Sklad (stav/prijem/inventura/pohyby/doobjednat) + dodavatelia - mirror
src/routes/warehouse.js."""
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import Response
from pydantic import BaseModel

from .. import db
from ..deps import require_login, require_role, set_flash
from ..errors import AppError
from ..services.alerts import export_csv, otvorene_alerty, posli_suhrn
from ..services.stock import korekcia, prijem, stock_state

router = APIRouter(dependencies=[Depends(require_login)], tags=["warehouse"])


def _sklad_data(q: str) -> dict:
    """Sklad ma jednu stranku so zalozkami - mirror zobrazSklad() v routes/warehouse.js."""
    polozky = db.all("""SELECT i.*, c.name AS cat_name FROM items i
                        LEFT JOIN categories c ON c.id = i.category_id
                        WHERE i.active = 1
                          AND (? = '' OR i.code LIKE ? OR i.name LIKE ? OR i.location LIKE ?)
                        ORDER BY i.code""", (q, f"%{q}%", f"%{q}%", f"%{q}%"))
    for p in polozky:
        p["stav"] = stock_state(p)

    pohyby = db.all("""SELECT m.*, i.code, i.name, i.unit, u.full_name,
                               rq.number AS ziadanka
                        FROM stock_moves m JOIN items i ON i.id = m.item_id
                        LEFT JOIN users u ON u.id = m.user_id
                        LEFT JOIN request_lines rl ON rl.id = m.request_line_id
                        LEFT JOIN requests rq ON rq.id = rl.request_id
                        ORDER BY m.id DESC LIMIT 100""")

    return {
        "q": q, "polozky": polozky, "pohyby": pohyby,
        "alerty": otvorene_alerty(),
        "dodavatelia": db.all("SELECT id, name FROM suppliers WHERE active=1 ORDER BY name"),
    }


@router.get("/sklad")
async def sklad(tab: str = "stav", q: str = "", user: dict = Depends(require_role("storekeeper", "buyer"))):
    povolene = ("stav", "prijem", "inventura", "pohyby")
    return {"zalozka": tab if tab in povolene else "stav", **_sklad_data(q.strip())}


@router.get("/sklad/doobjednat")
async def sklad_doobjednat(q: str = "", user: dict = Depends(require_role("storekeeper", "buyer"))):
    return {"zalozka": "doobjednat", **_sklad_data(q.strip())}


class PrijemBody(BaseModel):
    item_id: int
    qty: float
    supplier_id: Optional[int] = None
    note: Optional[str] = None


@router.post("/sklad/prijem")
async def sklad_prijem(request: Request, body: PrijemBody, user: dict = Depends(require_role("storekeeper"))):
    try:
        prijem(body.item_id, body.qty, body.supplier_id, user["id"], body.note)
        set_flash(request, "Príjem zaevidovaný.")
    except AppError as e:
        set_flash(request, str(e))
    return {"ok": True, "redirect": "/sklad?tab=prijem"}


class KorekciaBody(BaseModel):
    item_id: int
    new_qty: float
    note: Optional[str] = None
    typ: str = "INVENTURA"


@router.post("/sklad/korekcia")
async def sklad_korekcia(request: Request, body: KorekciaBody, user: dict = Depends(require_role("storekeeper"))):
    try:
        korekcia(body.item_id, body.new_qty, user["id"], body.note,
                "KOREKCIA" if body.typ == "KOREKCIA" else "INVENTURA")
        set_flash(request, "Stav upravený.")
    except AppError as e:
        set_flash(request, str(e))
    return {"ok": True, "redirect": "/sklad?tab=inventura"}


@router.post("/sklad/doobjednat/suhrn")
async def sklad_doobjednat_suhrn(request: Request, user: dict = Depends(require_role("storekeeper", "buyer"))):
    n = posli_suhrn()
    set_flash(request, f"Podklad na doobjednanie odoslaný ({n} položiek)." if n
              else "Žiadne položky pod signálnou zásobou.")
    return {"ok": True, "redirect": "/sklad/doobjednat"}


@router.get("/sklad/doobjednat/export.csv")
async def sklad_export_csv(user: dict = Depends(require_role("storekeeper", "buyer"))):
    from datetime import date
    datum = date.today().isoformat()
    return Response(
        content=export_csv(), media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="doobjednat-{datum}.csv"'})


# Stare adresy nech nekoncia na 404.
@router.get("/upozornenia")
async def upozornenia_redirect():
    return {"redirect": "/sklad/doobjednat"}


@router.get("/sklad/karta/{item_id}")
async def sklad_karta_redirect(item_id: int):
    return {"redirect": f"/produkt/{item_id}"}


# ---------------- DODAVATELIA ----------------

@router.get("/dodavatelia")
async def dodavatelia():
    zoznam = db.all("""SELECT s.*,
        (SELECT COUNT(*) FROM supplier_items si WHERE si.supplier_id = s.id) AS pocet_poloziek
        FROM suppliers s WHERE s.active = 1 ORDER BY s.name""")
    return {"dodavatelia": zoznam}


@router.get("/dodavatelia/{supplier_id}")
async def dodavatel_detail(supplier_id: int):
    d = db.get("SELECT * FROM suppliers WHERE id=?", (supplier_id,))
    if not d:
        raise HTTPException(status_code=404, detail="Dodávateľ neexistuje.")
    polozky = db.all("""SELECT i.*, si.supplier_sku, si.price, si.min_order_qty, si.is_primary
                        FROM supplier_items si JOIN items i ON i.id = si.item_id
                        WHERE si.supplier_id = ? ORDER BY i.code""", (supplier_id,))
    for p in polozky:
        p["stav"] = stock_state(p)
    ponuky = db.all("""SELECT q.*,
        (SELECT COUNT(*) FROM quote_lines ql WHERE ql.quote_id = q.id) AS riadkov
        FROM quotes q WHERE q.supplier_id = ? ORDER BY q.issued_at DESC""", (supplier_id,))
    return {"d": d, "polozky": polozky, "ponuky": ponuky}
